src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AggregatedDocumentDiagnosticSource.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+/// <summary>
+/// Aggregates multiple <see cref="IDiagnosticSource"/>s into a single source.
+/// </summary>
+/// <param name="sources">Sources to aggregate</param>
+/// <remarks>
+/// Aggregation is usually needed for clients like VS Code which supports single source per request.
+/// </remarks>
+internal sealed class AggregatedDocumentDiagnosticSource(ImmutableArray<IDiagnosticSource> sources) : IDiagnosticSource
+{
+    public static ImmutableArray<IDiagnosticSource> AggregateIfNeeded(IEnumerable<IDiagnosticSource> sources)
+    {
+        var result = sources.ToImmutableArray();
+        if (result.Length > 1)
+        {
+            result = [new AggregatedDocumentDiagnosticSource(result)];
+        }
+
+        return result;
+    }
+
+    public bool IsLiveSource() => true;
```

Comment:
wasn't there a check that they wwre all live sources to begin with?

Reply:
ah, I should have kept this class as private